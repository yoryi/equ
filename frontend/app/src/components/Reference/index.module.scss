@import 'src/styles/mixins/breakpoints';

:local(.container) {
    flex: 1;
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: #ffffff;
    border-radius: 12px;
    box-shadow: 0 4px 8px rgba(0, 29, 64, 0.04), 0 0 32px rgba(0, 29, 64, 0.04);
    padding: 32px;

    @include breakpoint('medium') {
        align-items: flex-start;
    }

    :local(.profileContainer) {
        display: flex;
        flex-direction: column;
        width: 100%;

        @include breakpoint('medium') {
            flex-direction: row;
            align-items: center;
        }

        :local(.avatar) {
            width: 64px;
            height: 64px;
            border-radius: 50%;
            align-self: center;

            &:local(.empty) {
                display: flex;
                justify-content: center;
                align-items: center;
                background-color: var(--equedi-blue-20);

                :local(.initials) {
                    font-family: Telegraf, sans-serif;
                    font-size: 24px;
                    line-height: 24px;
                    margin-top: 4px;
                    font-weight: bold;
                    color: var(--main-dark);
                    text-align: center;
                }
            }
        }

        :local(.detailsContainer) {
            display: flex;
            flex-direction: column;
            justify-content: center;
            flex: 1 1 0;
            min-width: 0;
            text-align: center;

            @include breakpoint('medium') {
                justify-content: flex-start;
                text-align: left;
                margin-left: 16px;
            }

            :local(.name) {
                margin-top: 8px;
                font-family: Ilisarniq, sans-serif;
                color: var(--main-dark);
                font-size: 24px;
                line-height: 32px;
                font-weight: bold;
                letter-spacing: -0.04em;
                overflow-wrap: break-word;
                text-overflow: ellipsis;
                white-space: nowrap;
                overflow: hidden;

                @include breakpoint('medium') {
                    margin-top: 0;
                    font-size: 20px;
                    line-height: 24px;
                }
            }

            :local(.position) {
                font-family: Ilisarniq, sans-serif;
                color: var(--grey);
                font-size: 21px;
                line-height: 24px;
                font-weight: bold;
                letter-spacing: -0.04em;

                @include breakpoint('medium') {
                    font-size: 16px;
                    line-height: 24px;
                    font-weight: normal;
                }
            }

            :local(.classes) {
                font-family: Ilisarniq, sans-serif;
                color: var(--grey);
                font-size: 16px;
                line-height: 24px;

                @include breakpoint('medium') {
                    font-size: 16px;
                    line-height: 24px;
                }
            }
        }
    }

    :local(.button) {
        margin-top: 16px;

        @include breakpoint('medium') {
            align-self: center;
        }
    }

    :local(.seeRelevantExperience) {
        margin-top: 24px;
        font-family: Ilisarniq, sans-serif;
        font-size: 14px;
        line-height: 16px;
        color: var(--main-blue);

        @include breakpoint('small') {
            width: 100%;
            text-align: center;
        }
    }
}
