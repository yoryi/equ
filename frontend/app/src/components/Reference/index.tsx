//Utils
import classNames from "classnames"
import * as React from "react"
import { useTranslation } from "react-i18next"
//Hooks
import { useSelector } from "react-redux"
import { useParams } from "react-router"
//Components
import { HashLink } from "react-router-hash-link"

//Types
import {
  ActivityType,
  ReduxState,
  Reference as TReference,
} from "../../store/types"
import { nameAbbreviation } from "../../utils/nameAbbreviation"
import Button from "../Button/Button"
//Styles
import Styles from "./index.module.scss"

interface ReferenceProps {
  reference: TReference
  className?: string
  onClick?: () => void
}

const Reference: React.FC<ReferenceProps> = ({
  reference,
  className,
  onClick,
}) => {
  const params = useParams<{ id?: string }>()
  const id = params.id ? parseInt(params.id) : null

  const mediaToken = useSelector<ReduxState, string | null>(
    (state) => state.media.mediaToken,
  )

  const { t } = useTranslation()

  let activityRoute: string
  switch (reference.referenceLink.activity?.type) {
    case ActivityType.SPORTS_SPIRIT:
    case ActivityType.ACADEMIC_ART:
    case ActivityType.HOBBIES: {
      activityRoute = `extracurriculars`
      break
    }
    case ActivityType.INTERNSHIP:
    case ActivityType.EMPLOYMENT: {
      activityRoute = `professional`
      break
    }
    default: {
      activityRoute = `service`
      break
    }
  }

  return (
    <div className={classNames(Styles.container, className)}>
      <div className={Styles.profileContainer}>
        {reference.media && (
          <img
            className={Styles.avatar}
            src={`${process.env.REACT_APP_API_URL}/media/${reference.media.path}?eqmt=${mediaToken}`}
            alt=""
          />
        )}

        {!reference.media && (
          <div className={classNames([Styles.avatar, Styles.empty])}>
            <span className={Styles.initials}>
              {nameAbbreviation(`${reference.firstName} ${reference.lastName}`)}
            </span>
          </div>
        )}

        <div className={Styles.detailsContainer}>
          <span className={Styles.name}>
            {`${reference.firstName} ${reference.lastName}`}
          </span>

          <span className={Styles.position}>{reference.position}</span>

          <span className={Styles.classes}>{reference.organization}</span>
        </div>
      </div>

      <Button className={Styles.button} size={`sm`} onClick={onClick}>
        {t(`common:readReference`)}
      </Button>

      {reference.referenceLink.activity && (
        <HashLink
          className={Styles.seeRelevantExperience}
          to={`${id ? `/student/${id}` : ``}/${activityRoute}#activity-${
            reference.referenceLink.activity.id
          }`}
          role={`activityLink`}
        >
          {t(`common:seeRelevantExperience`)}
        </HashLink>
      )}
    </div>
  )
}

export default Reference
