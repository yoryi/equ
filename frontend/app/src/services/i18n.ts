import filesize from "filesize"
import i18n from "i18next"
import _ from "lodash"
import { initReactI18next } from "react-i18next"

import { en } from "../translations"

void i18n.use(initReactI18next).init({
  ns: [`common`, `screens`],
  defaultNS: `screens`,
  resources: {
    en,
  },
  lng: `en`,
  fallbackLng: `en`,
  interpolation: {
    format: (value, format, lng) => {
      if (!format) {
        return value
      }

      return format.split(`,`).reduce((value: string, format: string) => {
        switch (format.trim()) {
          case `localeString`:
            return value.toLocaleString()
          case `enIndefiniteArticle`:
          case `enIndefiniteArticleCapitalized`: {
            if (!lng || lng !== `en`) {
              return value
            }

            const letters = [`a`, `e`, `i`, `o`, `u`, `h`]
            const firstLetter = value.substring(0, 1)
            let article = ``
            if (
              letters.find((letter) => firstLetter.toLowerCase() === letter)
            ) {
              article =
                format === `enIndefiniteArticleCapitalized` ? `An` : `an`
            } else {
              article = format === `enIndefiniteArticleCapitalized` ? `A` : `a`
            }

            return `${article} ${value}`
          }
          case `title`: {
            return value
              .toLowerCase()
              .split(` `)
              .map((word: string) => _.upperFirst(word))
              .join(` `)
          }
          case `lowerCase`: {
            return value.toLowerCase()
          }
          case `fileSize`: {
            return filesize(parseInt(value), { round: 0 })
          }
          default:
            return value
        }
      }, value)
    },
  },
})

export default i18n
