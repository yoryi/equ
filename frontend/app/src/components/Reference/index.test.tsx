//Testing
import { render } from "@testing-library/react"
import faker from "faker"
import * as React from "react"
//Translations
import { I18nextProvider } from "react-i18next"
//Redux store
import { Provider } from "react-redux"
//Navigation
import { MemoryRouter, Route } from "react-router"
import configureStore from "redux-mock-store"

import i18n from "../../services/i18n"
//Types
import {
  ActivityType,
  ReduxState,
  Reference as TReference,
  ReferenceType,
} from "../../store/types"
//Components
import Reference from "./"

type TestStore = Pick<ReduxState, "media">

describe(`Reference component`, () => {
  faker.seed(1000)

  const mockStore = configureStore<TestStore>()
  const store = mockStore({
    media: {
      mediaToken: faker.random.alphaNumeric(32),
    },
  })

  let initialEntries: string[] | null = null
  let reference: TReference
  beforeEach(() => {
    reference = {
      id: faker.random.number(10000),
      type: ReferenceType.Academic,
      firstName: faker.name.firstName(),
      lastName: faker.name.lastName(),
      media: {
        id: faker.random.number(10000),
        path: faker.image.imageUrl(),
      },
      organization: faker.company.companyName(),
      position: faker.lorem.word(),
      content: faker.lorem.sentences(3),
      isVisible: true,
      referenceLink: {
        id: faker.random.number(10000),
        type: ReferenceType.Academic,
        activity: null,
        token: faker.random.alphaNumeric(32),
        isActive: false,
        createdAt: null,
      },
    }
  })

  afterEach(() => {
    initialEntries = null
  })

  const wrapper: React.FC = ({ children }) => (
    <I18nextProvider i18n={i18n}>
      <Provider store={store}>
        <MemoryRouter initialEntries={initialEntries ?? [`/`]}>
          <Route path={[`/`, `/student/:id`]} exact>
            {children}
          </Route>
        </MemoryRouter>
      </Provider>
    </I18nextProvider>
  )

  describe(`should render and match the snapshot`, () => {
    describe(`media`, () => {
      it(`present`, () => {
        const { container } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`not present`, () => {
        reference = {
          ...reference,
          media: null,
        }

        const { container } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    it(`with activity`, () => {
      reference = {
        ...reference,
        type: ReferenceType.Activity,
        referenceLink: {
          ...reference.referenceLink,
          activity: {
            id: faker.random.number(10000),
            title: faker.lorem.words(2),
            startDate: null,
            endDate: null,
            company: faker.company.companyName(),
            description: faker.lorem.sentences(3),
            type: ActivityType.ACADEMIC_ART,
            referenceLink: null,
          },
        },
      }

      const { container } = render(<Reference reference={reference} />, {
        wrapper,
      })

      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe(`activity link`, () => {
    describe(`own profile`, () => {
      it(`SPORTS_SPIRIT`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.SPORTS_SPIRIT,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/extracurriculars#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`ACADEMIC_ART`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.ACADEMIC_ART,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/extracurriculars#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`HOBBIES`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.HOBBIES,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/extracurriculars#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`INTERNSHIP`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.INTERNSHIP,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/professional#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`EMPLOYMENT`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.EMPLOYMENT,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/professional#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`VOLUNTEER`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.VOLUNTEER,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/service#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`ACTIVISM`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.ACTIVISM,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/service#activity-${reference.referenceLink.activity?.id}`,
        )
      })
    })

    describe(`other student profile`, () => {
      let id: number
      beforeEach(() => {
        id = faker.random.number(10000)
        initialEntries = [`/student/${id}`]
      })

      it(`SPORTS_SPIRIT`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.SPORTS_SPIRIT,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/extracurriculars#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`ACADEMIC_ART`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.ACADEMIC_ART,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/extracurriculars#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`HOBBIES`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.HOBBIES,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/extracurriculars#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`INTERNSHIP`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.INTERNSHIP,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/professional#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`EMPLOYMENT`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.EMPLOYMENT,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/professional#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`VOLUNTEER`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.VOLUNTEER,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/service#activity-${reference.referenceLink.activity?.id}`,
        )
      })

      it(`ACTIVISM`, () => {
        reference = {
          ...reference,
          type: ReferenceType.Activity,
          referenceLink: {
            ...reference.referenceLink,
            activity: {
              id: faker.random.number(10000),
              title: faker.lorem.words(2),
              startDate: null,
              endDate: null,
              company: faker.company.companyName(),
              description: faker.lorem.sentences(3),
              type: ActivityType.ACTIVISM,
              referenceLink: null,
            },
          },
        }

        const { getByRole } = render(<Reference reference={reference} />, {
          wrapper,
        })

        expect(getByRole(`activityLink`)).toHaveAttribute(
          `href`,
          `/student/${id}/service#activity-${reference.referenceLink.activity?.id}`,
        )
      })
    })
  })
})
