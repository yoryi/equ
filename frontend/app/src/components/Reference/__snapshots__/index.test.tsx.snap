// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference component should render and match the snapshot media not present 1`] = `
<div
  class="container"
>
  <div
    class="profileContainer"
  >
    <div
      class="avatar empty"
    >
      <span
        class="initials"
      >
        MK
      </span>
    </div>
    <div
      class="detailsContainer"
    >
      <span
        class="name"
      >
        Maritza Kertzmann
      </span>
      <span
        class="position"
      >
        nam
      </span>
      <span
        class="classes"
      >
        Brakus - Hagenes
      </span>
    </div>
  </div>
  <button
    class="primary-button button default-button sm-button blue-button"
    style="min-width: 140px;"
  >
    Read Reference
  </button>
</div>
`;

exports[`Reference component should render and match the snapshot media present 1`] = `
<div
  class="container"
>
  <div
    class="profileContainer"
  >
    <img
      alt=""
      class="avatar"
      src="https://mock.equedi.com/media/http://lorempixel.com/640/480?eqmt=n74ly0hcvu701ce68uuk7rqwe260qw2t"
    />
    <div
      class="detailsContainer"
    >
      <span
        class="name"
      >
        Aryanna Wilderman
      </span>
      <span
        class="position"
      >
        doloremque
      </span>
      <span
        class="classes"
      >
        Flatley, Jerde and Okuneva
      </span>
    </div>
  </div>
  <button
    class="primary-button button default-button sm-button blue-button"
    style="min-width: 140px;"
  >
    Read Reference
  </button>
</div>
`;

exports[`Reference component should render and match the snapshot with activity 1`] = `
<div
  class="container"
>
  <div
    class="profileContainer"
  >
    <img
      alt=""
      class="avatar"
      src="https://mock.equedi.com/media/http://lorempixel.com/640/480?eqmt=n74ly0hcvu701ce68uuk7rqwe260qw2t"
    />
    <div
      class="detailsContainer"
    >
      <span
        class="name"
      >
        Waldo Kulas
      </span>
      <span
        class="position"
      >
        aliquam
      </span>
      <span
        class="classes"
      >
        Toy - Bauch
      </span>
    </div>
  </div>
  <button
    class="primary-button button default-button sm-button blue-button"
    style="min-width: 140px;"
  >
    Read Reference
  </button>
  <a
    class="seeRelevantExperience"
    href="/extracurriculars#activity-619"
    role="activityLink"
  >
    See Relevant Experience
  </a>
</div>
`;
